### ISSUE TITLE
Converting from dense to PyG format

### ISSUE THREAD

**Issue by legohyl** [pyg-team/pytorch_geometric#2093]
## ❓ Questions & Help

hey @rusty1s , I'm looking for some help on how to convert from a dense format `batch x num_nodes x latent_dim` to the required PyG `x` format. Context is, I wish to use `GATConv` as the first few layers of a model, and the last layer has a custom algorithm that only can work in the dense format. At the moment, my data is passed such that all samples have the same number of nodes (though I wish to extend this beyond that), and my `edge_index` is saved in the PyG format such that I have `batch x 2 x num_nodes*2`. All graphs are undirected here.

Is there a quick way to convert the dense format to the required input for the layer? Also, is there a way to handle if there are differing graph sizes?

**Comment by rusty1s** [comment]
```python
x = x.view(batch * num_nodes, latent_dim)
edge_index = edge_index.permute(1, 0, 2)
edge_index += torch.arange(batch).view(1, -1, 1) * batch
edge_index = edge_index.view(2, -1)
```
should do the trick.

The PyG format can handle different graph sizes, while the dense format cannot. You need to pad your node features with zeros in order to do so.

IMO, the best way to implement is to save your data in PyG format, and then utilize `torch_geometric.to_dense_batch` to convert your node features to format `[batch_size, num_nodes, latent_dim]`.

**Comment by legohyl** [comment]
Thanks @rusty1s ! I'm just curious, why is there a need to permute `edge_index` ? I'm guessing it's because of how the arrangement works in 3D tensors, do you have any links where I can read about how these work or how they are arranged?

I'd agree that saving the data in PyG format would be better. But here's the issue I'm facing, I'm actually working on a reinforcement learning problem, and each sample is a graph with features and the adjacency matrix. Eventually, I dump all of these samples into a large memory buffer, and sample a batch out every time. `node_features` are probably easy to handle because I can just concatenate them along `dim=0` to get the correct PyG format. It's handling the edges that I'm a bit unsure of; do I save these edges in their native adjacency format, or the `edge_index` sparse-like format first?

**Comment by legohyl** [comment]
Also, I think this `edge_index += torch.arange(batch).view(1, -1, 1) * batch` is supposed to be `edge_index += torch.arange(max_nodes).view(1, -1, 1) * max_nodes` right? Since each entry along `dim=0` `edge_index` is one sample. We should be incrementing them by the maximal number of nodes possible in one graph along `batch` axis.

**Comment by rusty1s** [comment]
Yes, it should be:
```
edge_index += torch.arange(batch).view(1, -1, 1) * max_num_nodes
```
I recommend storing them in the `edge_index` format, since it is really universal.

**Comment by legohyl** [comment]
Thanks @rusty1s ! After adding some of your suggestions and running the `GATConv`, I'm facing this issue with `torch`
```
/opt/conda/conda-bld/pytorch_1603729096996/work/aten/src/ATen/native/cuda/Indexing.cu:658: indexSelectLargeIndex: block: [57,0,0], thread: [21,0,0] Assertion `srcIndex < srcSelectDimSize` failed.
```
I'm not really sure what's happening here, illegal memory access?

**Comment by rusty1s** [comment]
Seems to be an out-of-bounds index error, so I suggest to check that `edge_index` does not refer to indices outside of `x`, i.e., `edge_index.max() < x.size(0)`.

**Comment by legohyl** [comment]
hey @rusty1s , i'm reopening this to ask a question similar to my above one. Now, I have a list of graphs where `[g1, g2, ...]` and `[e_index1, e_index2, ...]`, where the graphs are of different sizes and I have their corresponding PyG `edge_index`. How do I turn this into a valid batch? In this case, one list represents one batch.

**Comment by legohyl** [comment]
Rather, could you verify my implementation?
```
def list_convert_pyg(observation, var_adj, max_num_nodes):
    batch_size = len(observation)
    latent_dim = observation[0].size(1)
    
    x = torch.cat(observation, dim=0)
    for i, e in enumerate(var_adj):
        if i != 0:
            var_adj[i] = var_adj[i] + var_adj[i-1].max() + 1
    e_index = torch.cat(var_adj, dim=1)

    return x, e_index
```

**Comment by rusty1s** [comment]
Looks correct to me :)

**Comment by Boltzmachine** [comment]
> Rather, could you verify my implementation?
> 
> ```
> def list_convert_pyg(observation, var_adj, max_num_nodes):
>     batch_size = len(observation)
>     latent_dim = observation[0].size(1)
>     
>     x = torch.cat(observation, dim=0)
>     for i, e in enumerate(var_adj):
>         if i != 0:
>             var_adj[i] = var_adj[i] + var_adj[i-1].max() + 1
>     e_index = torch.cat(var_adj, dim=1)
> 
>     return x, e_index
> ```

I wonder that why do you use `var_adj[i-1].max()` ? I think I will use the number of nodes in the previous graph.

**Comment by legohyl** [comment]
@Boltzmachine I used that because `var_adj` is a variable that is already in PyG `edge_index` form. So the largest value is the number of nodes in the graph.
